core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -345,6 +345,11 @@ object SparkSubmit {
         System.exit(CLASS_NOT_FOUND_EXIT_STATUS)
     }
 
+    // SPARK-4170
+    if (classOf[scala.App].isAssignableFrom(mainClass)) {
```

Comment:
@srowen This is not working since the passed in mainClass is not that of the Scala `object` that extends `App`. For example, the mainClass is `Test` while the class of object is `Test$`

Reply:
`scala.App` is a trait, not an object, and there is no `scala.App$` class in the generated bytecode. Can you clarify?

Reply:
I mean user's application like

```scala
object Test extends App {
    // spark code 
}
```

In this case, `mainClass` is `Test` and `classOf[scala.App].isAssignableFrom(Test)` would be false since it's `Test$` that extends `scala.App`

Reply:
I see, but if you are trying to run an `object` then you'd specify class name `Test$` to begin with, I think? Normally an `App` would be a `class`.

Reply:
Actually, Scala will compile `object Test` into two Java classes, `class Test$ implements scala.App` and `class Test`.  We specify class name `Test` whose `main` method would invoke that of `Test$`.

Reply:
Hm, OK, what happens if you have `class Test extends App`?
In any event, if we do need to check both `Test` and `Test$`, that's fine, that's an easy change. I'd open a follow-up minor JIRA and make the change as a PR.

Reply:
It can't be `class` since `main` method must be static

Reply:
I think you have a point. Feel free to open a new JIRA / PR to change the test here. It's not a common use case but worth getting right.

Reply:
Thanks, filed https://issues.apache.org/jira/browse/SPARK-26977